### PR TITLE
Ghetto mode fold encoding

### LINKED ISSUE #49: Fold encoder
Should normalize the animation

Ideas:
- vertex displacement should be similar in every step
- ease-in ease-out animations for more natural feel
- configurable FPS

### PR DESCRIPTION
**Pull request by zemiret**
Closes #49

### CODE REVIEW

File: backend/origuide/fold/fold_encoder.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    @abc.abstractmethod
+    def next_step(self):
+        pass
+
+    @abc.abstractmethod
+    def advance(self):
+        pass
+
+
+class LogFoldEncoder(FoldEncoder):
```
Comment (maciekmm) [inline backend/origuide/fold/fold_encoder.py:18]: (y) nice.

File: backend/origuide/fold/fold_encoder.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import abc
+
+
+class FoldEncoder(abc.ABC):
```
Comment (maciekmm) [inline backend/origuide/fold/fold_encoder.py:4]: :+1: Nice

File: backend/origuide/fold/test_fold_producer.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import unittest
+
+from .fold_encoder import LogFoldEncoder
+
+
+class TestLogFoldEncoder(unittest.TestCase):
+    def test_next_step_resets_encoder(self):
+        encoder = LogFoldEncoder(4, 2, 1)
```
Comment (maciekmm) [inline backend/origuide/fold/test_fold_producer.py:8]: Personal preference:
I'd use named parameters for primitive types. `LogFoldEncoder(4, 2, 1)` is very cryptic and requires reader to navigate to the implementation.

`LogFoldEncoder(frame_drop_rate=4,
                 frame_drop_multiplier=2,
                 frame_drop_change_iter=1)` looks more descriptive

File: backend/origuide/main.py [hunk 4]
```diff
@@ -60,29 +60,23 @@ def create_faces(vertices, edges, faces_vertices):
...
                          fold.assignments)
 
     faces = create_faces(vertices, edges, fold.faces)
 
-    # TODO: Maybe some graph would be a more appropriate structure?
-    # IDEA: Create vertices, beams, etc "globally", and assign only their IDs to some more advanced objects
-    # that can extend the behavior
-
     for steady_state in fold.steady_states:
         # TODO: only assignments change between frames?
```
Comment (maciekmm) [inline backend/origuide/main.py:81]: and targetAngles

File: backend/origuide/fold/fold_producer.py [hunk 5]
```diff
@@ -27,16 +29,20 @@ def next_transition(self):
...
-            }
-        )
+    def accept(self, vertices: List[Vertex], last_step=False):
+        frame = {
+            "frame_inherit": True,
+            "frame_parent": len(self.transitions) - 1,
+            "vertices_coords": list(map(lambda x: list(x.pos.vec), vertices))
+        }
+        encoded = self.encoder.encode(frame, last_step)
+        self.encoder.advance()
```
Comment (maciekmm) [inline backend/origuide/fold/fold_producer.py:42]: Do we ever encode multiple times to advance? Maybe the logic could be hidden in .encode()?

File: backend/origuide/solver.py [hunk 6]
```diff
@@ -114,10 +115,6 @@ def _set_target_angles(self):
     def _total_forces_vecs(self):
         return list(map(lambda v: v.total_force().vec, self.vertices))
 
-    def _should_continue(self, prev_forces, cur_forces):
-        # TODO: Add iteration count also as a stop condition?
```
Comment (maciekmm) [inline backend/origuide/solver.py:118]: What about this TODO? Why have you given up on this idea?

**Review by maciekmm — APPROVED** [review]

File: backend/origuide/solver.py (hunk 6 above)
Comment (zemiret) [reply backend/origuide/solver.py:118]: It's out of scope of this feature.
And I don't know yet if I've given up on it. But in general, if iteration count is too high, then it means something is messed up, and potentially the model will never fold.

File: backend/origuide/fold/fold_producer.py (hunk 5 above)
Comment (zemiret) [reply backend/origuide/fold/fold_producer.py:42]: I wanted to split these 2 functions in interface, so as to allow skipping frames by calling advance, without calling encode. Could potentially prove useful, maybe.

File: backend/origuide/main.py (hunk 4 above)
Comment (zemiret) [reply backend/origuide/main.py:81]: I guess that event was your comment :)

File: backend/origuide/fold/test_fold_producer.py (hunk 3 above)
Comment (zemiret) [reply backend/origuide/fold/test_fold_producer.py:8]: I agree. I'll leave it like that in tests but change in main.

**Pull request merged**